### PR TITLE
[CodeGen] Add nneg and disjoint flags

### PR DESCRIPTION
**Pull request by tschuett**
MachineInstr learned the new flags.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-nneg-disjoint.ll [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+define i32 @call_not_nneg(i16 %a) {
+  ; CHECK-LABEL: name: call_not_nneg
+  ; CHECK: bb.1.entry:
+  ; CHECK-NEXT:   liveins: $w0
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $w0
+  ; CHECK-NEXT:   [[TRUNC:%[0-9]+]]:_(s16) = G_TRUNC [[COPY]](s32)
+  ; CHECK-NEXT:   %2:_(s32) = nneg G_ZEXT [[TRUNC]](s16)
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-nneg-disjoint.ll:26]: I am unsure about the nneg on this line.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 2]
```diff
@@ -1562,9 +1562,15 @@ bool IRTranslator::translateCast(unsigned Opcode, const User &U,
   if (U.getType()->getScalarType()->isBFloatTy() ||
       U.getOperand(0)->getType()->getScalarType()->isBFloatTy())
     return false;
+
+  uint32_t Flags = 0;
+  if (isa<Instruction>(U)) {
+    const Instruction &I = cast<Instruction>(U);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1568]: dyn_cast instead of isa+cast

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 3]
```diff
@@ -555,6 +555,16 @@ uint32_t MachineInstr::copyFlagsFromInstruction(const Instruction &I) {
       MIFlags |= MachineInstr::MIFlag::NoUWrap;
   }
 
+  // Copy the nonneg flag.
+  if (const PossiblyNonNegInst *PNI = dyn_cast<PossiblyNonNegInst>(&I))
+    if (PNI->hasNonNeg())
+      MIFlags |= MachineInstr::MIFlag::NonNeg;
+
+  // Copy the disjoint flag.
+  if (const PossiblyDisjointInst *PD = dyn_cast<PossiblyDisjointInst>(&I))
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineInstr.cpp:564]: else if? these are mutually exclusive?

Comment (tschuett) [reply llvm/lib/CodeGen/MachineInstr.cpp:564]: I followed the surround code style.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 2 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp]: Done.

File: llvm/lib/CodeGen/MachineInstr.cpp (hunk 3 above)
Comment (tschuett) [reply llvm/lib/CodeGen/MachineInstr.cpp:564]: I would prefer to keep it like this: for documentation and whenever the flags change again.

Comment (arsenm) [reply llvm/lib/CodeGen/MachineInstr.cpp:564]: The possible set of flag changes isn't unbounded

Comment (tschuett) [reply llvm/lib/CodeGen/MachineInstr.cpp]: Done.

File: llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-nneg-disjoint.ll [hunk 4]
```diff
@@ -0,0 +1,77 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=aarch64-linux-gnu -O0 -stop-after=irtranslator -global-isel -verify-machineinstrs %s -o - 2>&1 | FileCheck %s
+
+define i32 @call_nneg(i16 %a) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-nneg-disjoint.ll:4]: Test vector cases

**Review by arsenm — APPROVED** [review]
Could use vector tests

File: llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-nneg-disjoint.ll (hunk 4 above)
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-nneg-disjoint.ll:4]: Done.

**Pull request merged**